reconcile/aws_saml_roles/integration.py

```diff
@@ -0,0 +1,160 @@
+):
+    """Manage the SAML IAM roles for all AWS accounts with SSO enabled"""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_early_exit_desired_state(
+        self, query_func: Callable | None = None
+    ) -> dict[str, Any]:
+        """Return the desired state for early exit."""
+        if not query_func:
+            query_func = gql.get_api().query
+        return {
+            "aws_groups": [c.dict() for c in self.get_aws_groups(query_func)],
+        }
+
+    def get_aws_accounts(
+        self, query_func: Callable, account_name: str | None = None
+    ) -> list[AWSAccountV1]:
+        """Get all AWS accounts."""
+        data = aws_accounts_query(query_func)
+        return [
+            account
+            for account in data.accounts or []
+            if integration_is_enabled(self.name, account)
+            and (not account_name or account.name == account_name)
+        ]
+
+    def get_aws_groups(
+        self, query_func: Callable, account_name: str | None = None
+    ) -> list[AWSGroupV1]:
+        """Get all AWS groups with SSO enabled."""
+        data = aws_groups_query(query_func)
+        return [
+            group
+            for group in data.aws_groups or []
+            if integration_is_enabled(self.name, group.account)
+            and (not account_name or group.account.name == account_name)
+            and group.account.sso
+            and group.roles
+            and group.policies
+            and any(role.users for role in group.roles)
+        ]
+
+    @defer
+    def run(self, dry_run: bool, defer: Callable | None = None) -> None:
+        """Run the integration."""
+        gql_api = gql.get_api()
+        settings = queries.get_app_interface_settings()
+        aws_accounts = self.get_aws_accounts(
+            gql_api.query, account_name=self.params.account_name
+        )
+        aws_accounts_dict = [account.dict(by_alias=True) for account in aws_accounts]
+
+        ts = TerrascriptClient(
+            self.name.replace("-", "_"),
+            "",
+            self.params.thread_pool_size,
+            aws_accounts_dict,
+            settings=settings,
+        )
+
+        for group in self.get_aws_groups(
+            gql_api.query, account_name=self.params.account_name
+        ):
+            ts.populate_saml_iam_role(
+                account=group.account.name,
+                name=group.name,
+                saml_provider_name=self.params.saml_idp_name,
+                # aws groups without policies are filtered out by the query
+                policies=group.policies or [],
+                max_session_duration_hours=self.params.max_session_duration_hours,
+            )
+        working_dirs = ts.dump(print_to_file=self.params.print_to_file)
+
+        if self.params.print_to_file:
+            sys.exit(ExitCodes.SUCCESS)
+
+        aws_api = AWSApi(1, aws_accounts_dict, settings=settings, init_users=False)
```

Comment:
This could also work with secret_reader as arg to avoid settings https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/utils/aws_api.py#L99

Reply:
fixed